actix-prost-build/src/conversions.rs

```diff
@@ -0,0 +1,435 @@
+pub struct ConvertFieldOptions {
+    pub ty: Option<String>,
+    pub val_override: Option<String>,
+    pub required: bool,
+}
+
+#[derive(Default, Debug)]
+struct ConvertOptions {
+    fields: Vec<(String, ConvertFieldOptions)>,
+    extra: Vec<ExtraFieldOptions>,
+}
+
+impl TryFrom<(&DescriptorPool, &MessageDescriptor)> for ConvertOptions {
+    type Error = String;
+
+    fn try_from(
+        (descriptors, message): (&DescriptorPool, &MessageDescriptor),
+    ) -> Result<Self, Self::Error> {
+        let message_extension = descriptors
+            .get_message_by_name("google.protobuf.MessageOptions")
+            .ok_or("MessageOptions not found")?
+            .extensions()
+            .find(|ext| ext.name() == "extra_fields")
+            .unwrap();
+
+        let fields_extension = descriptors
+            .get_message_by_name("google.protobuf.FieldOptions")
+            .ok_or("FieldOptions not found")?
+            .extensions()
+            .find(|ext| ext.name() == "convert")
+            .unwrap();
+
+        let options = message.options();
+        let extra = options
+            .get_extension(&message_extension)
+            .as_list()
+            .unwrap()
+            .iter()
+            .map(|v| {
+                let m = v.as_message().unwrap();
+                ExtraFieldOptions::from(m)
+            })
+            .collect();
+
+        let fields = message
+            .fields()
+            .map(|f| {
+                let options = f.options();
+                let ext_val = options.get_extension(&fields_extension);
+                let ext_val = ext_val.as_message().unwrap();
+
+                let convert_options = ConvertFieldOptions::from(ext_val);
+
+                (String::from(f.name()), convert_options)
+            })
+            .collect();
+        Ok(Self { fields, extra })
+    }
+}
+
+impl From<&DynamicMessage> for ConvertFieldOptions {
+    fn from(value: &DynamicMessage) -> Self {
+        Self {
+            ty: get_string_field(value, "type"),
+            val_override: get_string_field(value, "override"),
+            required: match value.get_field_by_name("required") {
+                Some(v) => v.as_bool().unwrap(),
+                None => false,
+            },
+        }
+    }
+}
+
+impl From<&DynamicMessage> for ExtraFieldOptions {
+    fn from(value: &DynamicMessage) -> Self {
+        Self {
+            name: get_string_field(value, "name").unwrap(),
+            ty: get_string_field(value, "type").unwrap(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct ConversionsGenerator {
+    // Shared messages with ActixGenerator
+    pub messages: Rc<HashMap<String, syn::ItemStruct>>,
+    descriptors: DescriptorPool,
+    // Prefix for the Convert trait (could be static?)
+    convert_prefix: TokenStream,
+    // Track already processed messages and their impls in a simple bitmap
+    // to prevent duplicated code generation
+    processed_messages: HashMap<String, i32>,
```

Comment:
you don't seem to use the values anywhere, change it to `HashSet`?

Reply:
The value is bitmap stored as integer. Btw, I should make it smaller and unsized
https://github.com/blockscout/actix-prost/blob/aeadcecdf0cd26a20c3a04ae815357e662fc5a5a/actix-prost-build/src/conversions.rs#L200
https://github.com/blockscout/actix-prost/blob/aeadcecdf0cd26a20c3a04ae815357e662fc5a5a/actix-prost-build/src/conversions.rs#L217
https://github.com/blockscout/actix-prost/blob/aeadcecdf0cd26a20c3a04ae815357e662fc5a5a/actix-prost-build/src/conversions.rs#L240-L244

Reply:
Isn't it easier just to store `HashSet<(String, MessageType)>`?

Reply:
It depends :)
At first, I tried this HashSet approach, but then thought that a simple bitmap would be almost the same in terms of LoC. Also, with HashSet we need to make 2 lookups instead of 1